cpp/include/cuspatial/detail/join/quadtree_point_in_polygon.cuh

```diff
@@ -136,52 +142,67 @@ std::pair<rmm::device_uvector<IndexType>, rmm::device_uvector<IndexType>> quadtr
                          local_point_offsets.begin() + 1);
 
   // Ensure local point offsets starts at 0
-  IndexType init{0};
+  std::uint64_t init{0};
   local_point_offsets.set_element_async(0, init, stream);
 
   // The last element is the total number of points to test against any polygon.
   auto num_total_points = local_point_offsets.back_element(stream);
 
-  // Allocate the output polygon and point index pair vectors
-  rmm::device_uvector<IndexType> poly_indices(num_total_points, stream);
-  rmm::device_uvector<IndexType> point_indices(num_total_points, stream);
-
-  auto poly_and_point_indices =
-    thrust::make_zip_iterator(poly_indices.begin(), point_indices.begin());
+  // Process in chunks of *at most* 25% of the free available device memory to avoid OOM'ing
+  // when allocating additional output space.
+  auto max_chunk_size =
+    static_cast<std::uint64_t>(rmm::percent_of_free_device_memory(25) / (sizeof(IndexType) * 2));
 
-  // Enumerate the point X/Ys using the sorted `point_indices` (from quadtree construction)
-  auto point_xys_iter = thrust::make_permutation_iterator(points_first, point_indices_first);
+  auto num_chunks = std::uint64_t{num_total_points / max_chunk_size + 1};
 
-  // Compute the combination of polygon and point index pairs. For each polygon/quadrant pair,
-  // enumerate pairs of (poly_index, point_index) for each point in each quadrant.
-  //
-  // In Python pseudocode:
-  // ```
-  // pp_pairs = []
-  // for polygon, quadrant in pq_pairs:
-  //   for point in quadrant:
-  //     pp_pairs.append((polygon, point))
-  // ```
-  //
-  auto global_to_poly_and_point_indices = detail::make_counting_transform_iterator(
-    0,
-    detail::compute_poly_and_point_indices{quad_offsets_iter,
-                                           local_point_offsets.begin(),
-                                           local_point_offsets.end(),
-                                           poly_indices_first});
-
-  // Compute the number of intersections by removing (poly, point) pairs that don't intersect
-  auto num_intersections = thrust::distance(
-    poly_and_point_indices,
-    thrust::copy_if(rmm::exec_policy(stream),
-                    global_to_poly_and_point_indices,
-                    global_to_poly_and_point_indices + num_total_points,
-                    poly_and_point_indices,
-                    detail::test_poly_point_intersection{point_xys_iter, polygons}));
+  // Allocate the output polygon and point index pair vectors
+  rmm::device_uvector<IndexType> poly_indices(0, stream);
+  rmm::device_uvector<IndexType> point_indices(0, stream);
+
+  std::uint64_t num_intersections = 0;
```

Comment:
`num_intersections` isn't numerically required to be a uint64 (since we use uint32 for the output index cols), but using this type avoids needing to cast in the `chunk_size` min calc on line 165.